pkg/server/test/list_objects.go

```diff
@@ -14,460 +12,205 @@ import (
-				request:        newListObjectsRequest(store, "repo", "unknown", "user:anna", model.Id, nil),
-				expectedResult: nil,
-				expectedError:  serverErrors.RelationNotFound("unknown", "repo", nil),
-			},
-		}
-
-		typesys := typesystem.New(model)
-
-		connectedObjectsCmd := commands.ConnectedObjectsCommand{
-			Datastore:        ds,
-			Typesystem:       typesys,
-			ResolveNodeLimit: defaultResolveNodeLimit,
-			Limit:            defaultListObjectsMaxResults,
-		}
-
-		listObjectsQuery := &commands.ListObjectsQuery{
-			Datastore:             ds,
-			Logger:                logger.NewNoopLogger(),
-			ListObjectsDeadline:   defaultListObjectsDeadline,
-			ListObjectsMaxResults: defaultListObjectsMaxResults,
-			ResolveNodeLimit:      defaultResolveNodeLimit,
-			ConnectedObjects:      connectedObjectsCmd.StreamedConnectedObjects,
-			CheckResolver:         graph.NewLocalChecker(storage.NewContextualTupleDatastore(ds), checkConcurrencyLimit),
-		}
-
-		ctx = typesystem.ContextWithTypesystem(ctx, typesys)
-		runListObjectsTests(t, ctx, testCases, listObjectsQuery)
-	})
-}
-
 type mockStreamServer struct {
 	grpc.ServerStream
 	channel chan string
 }
 
-func NewMockStreamServer(size int) *mockStreamServer {
-	return &mockStreamServer{
-		channel: make(chan string, size),
-	}
-}
-
 func (x *mockStreamServer) Send(m *openfgapb.StreamedListObjectsResponse) error {
 	x.channel <- m.Object
 	return nil
 }
 
-func runListObjectsTests(t *testing.T, ctx context.Context, testCases []listObjectsTestCase, listObjectsQuery *commands.ListObjectsQuery) {
-
-	for _, test := range testCases {
-		ctx = storage.ContextWithContextualTuples(ctx, test.request.ContextualTuples.GetTupleKeys())
-
-		t.Run(test.name+"/streaming", func(t *testing.T) {
-			server := NewMockStreamServer(len(test.expectedResult))
-
-			done := make(chan struct{})
-			var streamedObjectIds []string
-			go func() {
-				for x := range server.channel {
-					streamedObjectIds = append(streamedObjectIds, x)
-				}
-
-				done <- struct{}{}
-			}()
-
-			err := listObjectsQuery.ExecuteStreamed(ctx, &openfgapb.StreamedListObjectsRequest{
-				StoreId:              test.request.StoreId,
-				AuthorizationModelId: test.request.AuthorizationModelId,
-				Type:                 test.request.Type,
-				Relation:             test.request.Relation,
-				User:                 test.request.User,
-				ContextualTuples:     test.request.ContextualTuples,
-			}, server)
-
-			close(server.channel)
-			<-done
+type listObjectsTestCase struct {
+	name                   string
+	schema                 string
+	tuples                 []*openfgapb.TupleKey
+	model                  string
+	objectType             string
+	user                   string
+	relation               string
+	contextualTuples       *openfgapb.ContextualTupleKeys
+	allResults             []string //all the results. the server may return less
+	maxResults             uint32
+	minimumResultsExpected uint32
```

Comment:
> I don't see why the added complexity for minimumResultsExpected is necessary

There has been more than one time where, while writing the test cases, I had introduced a problem in the setup and the LO call returned 0 objects and i could not understand why the test was passing

> expected should be a subset of the actual 

If `expected` has length 0 this assertion will pass

> and the len(expected) == maxResults

No because if I expect only one result and `maxResults=2` (i'll add a test for this), this assertion will fail. The correct assertion would be `len(expected) <= maxResults`. So i'm only checking on the upper bound, I want to add an assertion for the lower bound too.

Reply:
If you expect, for example, `["document:a", "document:b"]` and `maxResults=3`, then why does it not suffice to write something like:

```go
var res *openfgapb.ListObjectsResponse
var err error

require.NotNil(t, res)
require.NoError(t, err)
require.LessOrEqual(t, len(res.Objects), int(test.maxResults))
require.Equal(t, len(test.Expected), len(res.Objects)) 
require.ElementsMatch(t, test.Expected, res.Objects)
```

You assert that the response has `<= maxResults` && the number of returned objects is equal to the expected number && the elements match (regardless of order). This would have caught your failures as well, because the `len(t.Expected)` would have been greater than the `len(res.Objects)`.

This seems more readable and cleaner than introducing a `mininumResultsExpected` variable. Is there a reason this wouldn't work?

Reply:
How would I write `test.Expected`? Say all the valid responses are `document:1, document:2, document:3`. I ask for two of those. But the code may give me `doc:1, doc:2` or `doc:2, doc:3` or `doc:1,doc:3`.

I'd have to write every possible permutation in the test setup, which is doable but more cumbersome than what I propose here

Reply:
> Say all the valid responses are document:1, document:2, document:3. I ask for two of those. But the code may give me doc:1, doc:2 or doc:2, doc:3 or doc:1,doc:3.

Gotcha, I see your point. If you write, for example,

`doc:1#viewer@user:jon`
`doc:2#viewer@user:jon`
`doc:3#viewer@user:jon`

and you are testing `ListObjects(doc#viewer, user:jon)` with `maxResults=2`, then your `test.Expected` slice would be `[doc:1, doc:2, doc:3]`. For example,

```go
expected := []string{"doc:1", "doc:2", "doc:3"}

var res *openfgapb.ListObjectsResponse
var err error

require.NotNil(t, res)
require.NoError(t, err)

actual := res.Objects
require.LessOrEqual(t, len(actual), int(test.maxResults))

if len(expected) > 0 {
	require.NotEmpty(t, actual) // fails if actual=[]string{}
}

require.Subset(t, expected, actual) // fails unless all elements in 'actual' are in 'expected'
```

Reply:
```golang
expected := []string{"doc:1", "doc:2", "doc:3"}
```

If I do `ListObjects(doc#viewer, user:jon, maxResults=2)` and I only got back `doc:1`, your suggestion will pass but mine will fail, which is what I want

Reply:
Gotcha 👍 .